src/components/verification-grid/verification-grid.ts

```diff
@@ -292,6 +338,194 @@ export class VerificationGridComponent extends AbstractComponent(LitElement) {
+
+    const distance = Math.abs(targetAspectRatio - candidateSizeAspectRatio);
+    const maxRatio = Math.max(targetAspectRatio, candidateSizeAspectRatio);
+    const similarity = 1 - distance / maxRatio;
+
+    return similarity;
+  }
+
+  private computeGridShape(target: number): GridShape {
+    if (target === 1) {
+      return { columns: 1, rows: 1 };
+    } else if (target === 4) {
+      // TODO: Find out why 2x2 is not being preferred in our current algorithm
+      // and fix it so that we don't have to use this hack
+      //
+      // our current implementation works well for most grid sizes
+      // however, for a grid target of 4, it will prefer a 4x1 grid over a 2x2
+      // this is because 2x2 does not meet our aspect ratio threshold
+      // this is expected because we don't usually want a 1:1 aspect ratio
+      // for a 16:9 screen
+      // however, in the case of a target of 4, we want a 2x2 grid
+      // I've used this hack to force a 2x2 grid for a target of 4
+      // when the grid containers aspect ratio is close to 16:9
+      const screenSize = window.screen;
+      const aspectRatio = screenSize.width / screenSize.height;
+
+      // we allow for a bit of deviation from the exact 16:9 aspect ratio so
+      // similar aspect ratios also get the nicer grid size for a target of 4
+      const matchingAspectRatio = 16 / 9;
+      const matchingDeviation = 0.2;
+
+      if (Math.abs(aspectRatio - matchingAspectRatio) <= matchingDeviation) {
+        return { columns: 2, rows: 2 };
+      }
+    }
+
+    const containerSize = this.gridContainer.getBoundingClientRect();
+    let refinedTarget = target;
+    const testedTargets: number[] = [];
+
+    // some numbers e.g. 9 have one factor such as 3x3 which creates a aspect
+    // ratio of 1.00
+    // using a 3x3 grid for a 16:9 screen is not ideal because it will create
+    // a lot of dead space
+    // therefore, we have a threshold that we have to meet. If we do not meet
+    // the threshold, we keep increasing the target until we find a grid shape
+    // that meets the similarity percentage threshold
+    const targetThreshold = 0.5;
+
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      const candidateShapes: GridShape[] = [];
+
+      // sometimes we can get stuck in an infinite loop where increasing grid
+      // size will result in the grid not fitting, but decreasing the grid size
+      // will result in a grid that does not meet the aspect ratio threshold
+      // to resolve this, we keep a record of the targets that we have checked
+      // if we see the same one twice (we have increased, then decreased)
+      // we will decrement the grid size until we find a grid that fits
+      if (testedTargets.includes(refinedTarget)) {
+        refinedTarget -= 1;
+        continue;
+      }
+      testedTargets.push(refinedTarget);
+
+      let hasFittingCandidate = false;
+      for (let i = 1; i <= refinedTarget; i++) {
```

Comment:
~so you're counting up from a grid-size of 1, to the target to find out what will fit?~

you counting up numbers to determine factors. Why don't you just calculate the factors as a separate step and then iterate though those?

Also we can optimize this step a bit. If we have our factor pairs we can assume the most "square" size is the most optimal... so we start of at a square root, and then work backwards. so for 12 we check 4 * 3 first, and then there exists the potential not to have to check every pair.